exporters/failure/collector_pagerduty.py

```diff
@@ -0,0 +1,162 @@
+#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+#    License for the specific language governing permissions and limitations
+#    under the License.
+#
+
+import logging
+
+import requests
+from attrs import define, field
+
+from failure.collector_base import AbstractFailureCollector, TrackerIssue
+from pelorus.config import env_var_names, env_vars
+from pelorus.config.converters import comma_or_whitespace_separated
+from pelorus.config.log import REDACT, log
+from pelorus.errors import FailureProviderAuthenticationError
+from pelorus.timeutil import parse_assuming_utc, second_precision
+from pelorus.utils import TokenAuth, set_up_requests_session
+
+_DATETIME_FORMAT = "%Y-%m-%dT%H:%M:%SZ"
+
+
+@define(kw_only=True)
+class PagerdutyFailureCollector(AbstractFailureCollector):
+    """
+    PagerDuty implementation of a FailureCollector
+    """
+
+    token: str = field(
+        default="",
+        metadata=env_vars(*env_var_names.TOKEN) | log(REDACT),
+        repr=False,
+    )
+
+    tls_verify: bool = field(default=True)
+
+    session: requests.Session = field(factory=requests.Session, init=False)
+
+    incident_urgency: set[str] = field(
+        factory=set,
+        converter=comma_or_whitespace_separated(set),
+        metadata=env_vars("PAGERDUTY_URGENCY"),
+    )
+
+    incident_priority: set[str] = field(
+        factory=set,
+        converter=comma_or_whitespace_separated(set),
+        metadata=env_vars("PAGERDUTY_PRIORITY"),
+    )
+
+    def __attrs_post_init__(self):
+        # disable .netrc
+        self.session.trust_env = False
+
+        if self.token:
+            set_up_requests_session(
+                self.session,
+                self.tls_verify,
+                auth=TokenAuth(self.token, is_pagerduty=True),
+            )
+
+    def get_incidents(self) -> list[dict]:
+        logging.info("Getting incidents")
+        url = "https://api.pagerduty.com/incidents?date_range=all&limit=10000"
+        headers = {
+            "Accept": "application/vnd.pagerduty+json;version=2",
+        }
+
+        resp = self.session.get(url, headers=headers)
+        try:
+            resp.raise_for_status()
+            # TODO too much noise?
+            logging.debug("PagerDuty successfully returned %s", resp.text)
+            return resp.json()["incidents"]
+        except requests.HTTPError as error:
+            if resp.status_code == requests.codes.unauthorized:
+                raise FailureProviderAuthenticationError from error
+            raise
+
+    def filter_by_urgency(self, urgency: str) -> bool:
+        if not self.incident_urgency:
+            return True
+        return urgency in self.incident_urgency
+
+    def filter_by_priority(self, priority: str) -> bool:
+        if not self.incident_priority:
+            return True
+        try:
+            return priority["summary"] in self.incident_priority
```

Comment:
Did wrong here, it should be `Optional[Dict[str,str]]`, it will be a dict or None https://developer.pagerduty.com/api-reference/9d0b4b12e36f9-list-incidents

if is a dict, it will have; if it does not have, means it is None, and falls back to the null case